sklearn/naive_bayes.py

```diff
@@ -1528,3 +1541,600 @@ def _joint_log_likelihood(self, X):
+        return np.sum(all_jlls, axis=0) - (n_estimators - 1) * log_prior
+
+    def _validate_estimators(self, check_partial=False):
+        # Check if estimators have fit/partial_fit and joint log prob methods
+        # Validate estimator names via _BaseComposition._validate_names(self, names)
+        if not self.nb_estimators:
+            raise ValueError(
+                "A list of naive Bayes estimators must be provided "
+                "in the form [(name, nb_estimator, columns), ... ]."
+            )
+        names, estimators, _ = zip(*self.nb_estimators)
+        for e in estimators:
+            if (not check_partial) and (
+                not (hasattr(e, "fit") and hasattr(e, "predict_joint_log_proba"))
+            ):
+                raise TypeError(
+                    "Estimators must be naive Bayes estimators implementing "
+                    "`fit` and `predict_joint_log_proba` methods."
+                )
+            if check_partial and (
+                not (
+                    hasattr(e, "partial_fit") and hasattr(e, "predict_joint_log_proba")
+                )
+            ):
+                raise TypeError(
+                    "Estimators must be Naive Bayes estimators implementing "
+                    "`partial_fit` and `predict_joint_log_proba` methods."
+                )
+        self._validate_names(names)
+
+    def _validate_column_callables(self, X):
+        """
+        Convert callable column specifications and store into self._columns.
+
+        Empty-set columns do not enjoy any special treatment.
+        """
+        # Almost a verbatim copy of ColumnTransformer._validate_column_callables().
+        # Consider refactoring in the future.
+        # Unlike ColumnTransformer, this estimator does not need to output a
+        # dataframe or validate a the remainder, so _estimator_to_input_indices
+        # is not really needed, but retained for consistency with
+        # ColumnTransformer code.
+        all_columns = []
+        estimator_to_input_indices = {}
+        for name, _, columns in self.nb_estimators:
+            if callable(columns):
+                columns = columns(X)
+            all_columns.append(columns)
+            estimator_to_input_indices[name] = _get_column_indices(X, columns)
+        self._columns = all_columns
+        self._estimator_to_input_indices = estimator_to_input_indices
+
+    @property
+    def named_estimators_(self):
+        """Access the fitted naive Bayes subestimators by name.
+
+        Read-only attribute to access any estimators by given name.
+        Keys are estimators names and values are the fitted estimator
+        objects.
+        """
+        # Almost a verbatim copy of ColumnTransformer.named_transformers_
+        # Use Bunch object to improve autocomplete
+        return Bunch(**{name: e for name, e, _ in self.estimators_})
+
+    def _iter(self, *, fitted=False, replace_strings=False):
```

Comment:
`_iter` factorises skipping sub-estimators with empty column selections. This is like "drop" in a way.

I think we should allow empty column selection. First, it may come out unexpectedly from callable column selectors. Second, it makes trying different combinations easier. For example,
```
	>>> clf = ColumnwiseNB(estimators=[('mnb1', MultinomialNB(), [0, 1]),
	...                                ('mnb2', MultinomialNB(), [3, 4]),
	...                                ('gnb1', GaussianNB(), [5])])
``` 
is easier to change by hand to
```
	>>> clf = ColumnwiseNB(estimators=[('mnb1', MultinomialNB(), [0, 1]),
	...                                ('mnb2', MultinomialNB(), [3, 4]),
	...                                ('gnb1', GaussianNB(), [])])
``` 
without needing to remove the last tuple, especially if the columns are assigned to a variable elsewhere, e.g.:
```
	>>> clf = ColumnwiseNB(estimators=[('mnb1', MultinomialNB(), [0, 1]),
	...                                ('mnb2', MultinomialNB(), [3, 4]),
	...                                ('gnb1', GaussianNB(), gaussian_features)])
``` 
Similarly, if the user writes a piece of code to iterate over possible column combinations, they won't have to consider empty selection as an edge case (except when all sub-estimators have empty column sets).

Note that `nb_estimators_` attribute keeps an original uncloned unfitted sub-estimator instance whenever empty column sets occur. This is because we want `nb_estimators` and `nb_estimators_` have equal numbers of elements, and ColumnTransformer does the same thing. Thus we have to skip empty-column sub-estimators at predict too.